contracts/contracts/compensation/CompensationClaims.sol

```diff
@@ -0,0 +1,168 @@
+pragma solidity 0.5.11;
+
+import {
+    Initializable
+} from "@openzeppelin/upgrades/contracts/Initializable.sol";
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import { SafeERC20 } from "@openzeppelin/contracts/token/ERC20/SafeERC20.sol";
+import { SafeMath } from "@openzeppelin/contracts/math/SafeMath.sol";
+import { Governable } from "../governance/Governable.sol";
+
+/**
+ * @title Compensation Claims
+ * @author Origin Protocol Inc
+ * @dev Airdrop for ERC20 tokens.
+ *
+ *   Provides a coin airdrop with a verification period in which everyone
+ *   can check that all claims are correct before any actual funds are moved
+ *   to the contract.
+ *
+ *      - Users can claim funds during the claim period.
+ *
+ *      - The adjuster can set the amount of each user's claim,
+ *         but only when unlocked, and not during the claim period.
+ *
+ *      - The governor can unlock and lock the adjuster, outside the claim period.
+ *      - The governor can start the claim period, if it's not started.
+ *      - The governor can collect any remaining funds after the claim period is over.
+ *
+ *  Intended use sequence:
+ *
+ *   1. Governor unlocks the adjuster
+ *   2. Adjuster uploads claims
+ *   3. Governor locks the adjuster
+ *   4. Everyone verifies that the claim amounts and totals are correct
+ *   5. Payout funds are moved to the contract
+ *   6. The claim period starts
+ *   7. Users claim funds
+ *   8. The claim period ends
+ *   9. Governor can collect any remaing funds
+ *
+ */
+contract CompensationClaims is Governable {
+    address public adjuster;
+    address public token;
+    uint256 public end;
+    uint256 public totalClaims;
+    mapping(address => uint256) claims;
+    bool public isAdjusterLocked;
+
+    using SafeMath for uint256;
+
+    event Claim(address indexed recipient, uint256 amount);
+    event ClaimSet(address indexed recipient, uint256 amount);
+    event Start(uint256 end);
+    event Lock();
+    event Unlock();
+    event Collect(address indexed coin, uint256 amount);
+
+    constructor(address _token, address _adjuster) public onlyGovernor {
+        token = _token;
+        adjuster = _adjuster;
+        isAdjusterLocked = true;
+    }
+
+    function balanceOf(address _account) external view returns (uint256) {
+        return claims[_account];
+    }
+
+    function decimals() external view returns (uint8) {
+        return IERC20Decimals(token).decimals();
+    }
+
+    /* -- User -- */
+
+    function claim(address _recipient) external onlyInClaimPeriod nonReentrant {
```

Comment:
Is it OK that anyone can claim on behalf of another user? Can we just use msg.sender instead of passing the _recipient arg?

Reply:
I wondered about this. My thought was that if money was locked up in a smart contract, it might not be setup to call this contract as a msg.sender by itself.

That said, I'm open to just using msg.sender, and handling anything else outside the normal flow.